src/core/mac/mac.cpp

```diff
@@ -1869,15 +1915,27 @@ Error Mac::ProcessEnhAckSecurity(TxFrame &aTxFrame, RxFrame &aAckFrame)
 
     if (ackKeyId == (keyManager.GetCurrentKeySequence() & 0x7f))
     {
+#if OPENTHREAD_CONFIG_PSA_CRYPTO_ENABLE
+        macKeyRef = mLinks.GetSubMac().GetCurrentMacKeyRef();
+#else
         macKey = &mLinks.GetSubMac().GetCurrentMacKey();
+#endif
```

Comment:
I have a suggestion that may help simplify the code:
- We can define a common new type (`class`) `Mac::KeyRef`.
- The `KeyRef` implementation can have a `union` and `mType` which indicates whether it is a PSA key ref or a `const Key *` (somewhat similar to how `Mac::Address` can contain a short address, extended address, or even be empty ).
- We can then use it here and other places and I think won't need the `#if OPENTHREAD_CONFIG_PSA_CRYPTO_ENABLE` checks.

Reply:
@abtink I like this idea and we did initially discuss this approach. But we chose the other implementation because we were not sure if mbedTLS and PSA will both be supported concurrently or if the PSA will replace the existing mbedTLS implementation. 

Even with this idea, removing the `OPENTHREAD_CONFIG_PSA_CRYPTO_ENABLE` entirely will not be easy. We will still need these in lower level crypto APIs to call either PSA or mbedTLS modules. 

I do have an additional suggestion which might solve this.

If we choose to support mbedTLS and PSA concurrently, as you suggest, we can create common classes/structures wherever necessary, like the `KeyRef` class you suggested, eliminating the need to conditionally compile bulk of the code. But for lower level Crypto APIs, we can introduce a crypto platform abstraction. This way vendors can choose implement their own crypto modules. This will get rid of all the conditional compilation. I do have a similar abstraction implemented, it is named psa.h, but can be rename to crypto if needed.

Reply:
@abtink @LuDuda Any ideas for how we can approach mPskc and mMasterKey? Initially I thought of modifying the otMasterKey and otPskc definitions to have a union but we have a problem dereferencing the packed structures as they might not be aligned. So either we can add a `OT_TOOL_PACKED_ALIGNED` macros or add the key ref to the `MasterKey` and `Pskc` classes. Which ones do you guys recommend?

Reply:
@hemanth-silabs I think we should re-think whether the new definitions for `otMasterKey` or `otPskc` need to be packed or not.

I think the reason they are currently defined as packed is that we can use the same type to append/read (e.g. as part of TLV) directly to/from a message (a buffer). I think in the case of TLV use,  we always need to literal key (16 bytes) to be appended and we never just append a PSA key ref (that's only for use within same devcie).

Here is my suggestion:
- New definitions of a `otMasterKey` and `otPskc` can use `union` style (to allow it to be either literal or a reference).
- They don't need to be packed.
- We need to handle the uses in the code when they are used in a TLV (to be read or appended in a message) seperately (since we need to use literal key in  these cases).

Reply:
The other option is to get the objects and the stack using the platform code to export the key to for the TLV.

Reply:
Here is my suggestion on this:
- We have a single method `KeyManager::GetMasterKey()` (not different overloads) which returns an object of type `const MasterKey &`.
- The new `MasterKey` class type itself will use `union` style (can store literal key or a PSA key ref).
- The new `MasterKey` will no longer be defined as "packed".
- We add methods in the new `MasterKey` class that let us retrieve the literal key (i.e., copy the key bytes into some buffer), like `void MasterKey::CopyKey(uint8_t *aBuffer)`
- This new method can be used when the literal key need to be included in message (in TLVs). 
- Same model with `Pskc`.

Reply:
But is the `Pskc` and `MasterKey` class available for all the modules? I am thinking about CLI modules. Wont this mean including key_manager header into all these modules?

Reply:
The CLI (and NCP) modules generally only see and use the public `C` OT API definitions (which are all headers/definitions in `include/openthread`). We don't want them to see or access any of the core `C++` definitions. 

`MasterKey` and `Pskc` are intended for internal (to core) use. There are similar public definitions `otMasterKey` and `otPskc` which can be used by CLI/NCP and in general OT API users.  Note that the public and internal types don't need to be exatcly the same (we can have ways to convert from one to other).  We probably need to think whether the user of OT API need to be aware whether PSA is being used or not (e.g., when user gets/sets the `otOperationalDataset` which includes a `otMasterKey mMasterKey` member field , do we want to provide the literal key to caller or PSA ref to key?).